### PR TITLE
Add type tagging support for F types to make tagging Seq, Future etc easier

### PR DESCRIPTION
**Pull request by borissmidt**
We can lift our matching to make it easier to tag an Seq of something.

### CODE REVIEW

**Comment by borissmidt** [comment]
btw i just found this maybe it is usefull to add it to the description for if you need some stronger type tagging
https://github.com/Rudogma/scala-supertagged

File: tagging/src/main/scala/com/softwaremill/tagging/package.scala [hunk 1]
```diff
@@ -23,12 +25,38 @@ package com.softwaremill
...
   implicit class Tagger[T](val t: T) extends AnyVal {
-    def taggedWith[U]: T @@ U = t.asInstanceOf[T @@ U]
+    @inline def taggedWith[U]: T @@ U = t.asInstanceOf[T @@ U]
   }
   implicit class AndTagger[T, U](val t: T @@ U) extends AnyVal {
-    def andTaggedWith[V]: T @@ (U with V) = t.asInstanceOf[T @@ (U with V)]
+    @inline def andTaggedWith[V]: T @@ (U with V) = t.asInstanceOf[T @@ (U with V)]
+
+    @inline def replaceTag[V]: T @@ V = t.asInstanceOf[T @@ V]
+    @inline def eraseTag: T = t.asInstanceOf[T]
```
Comment (adamw) [inline tagging/src/main/scala/com/softwaremill/tagging/package.scala:38]: isn't this just upcasting?

Comment (adamw) [reply tagging/src/main/scala/com/softwaremill/tagging/package.scala:38]: ah might make a difference in collections

**Comment by adamw** [comment]
Thanks! Should be released soon as 2.3.0

**Comment by borissmidt** [comment]
Oke, I've found another super cool tagging library that has more features and a little bit nicer syntax.
Maybe it is interesting to you:
The nice thing is that it is almost the same syntax as value classes and you can nicely scope your Ops implicit class Also they have a taggWithL which tags the bottom type.

https://github.com/Tvaroh/tagged-types

**Comment by adamw** [comment]
Yes there's a couple of alternatives, still I like the lightness of the above, covers most use-cases :) Thanks for the link

**Comment by borissmidt** [comment]
I wonder if these tagged types will still be possible in dotty.
it would have been cool if you had a special `opaque` type which which would have the operations of the original type.

**Comment by adamw** [comment]
I haven't tried, but I think this should compile using Dotty.

But as you write, most probably `opaque` types will replace tags in many usages.

**Comment by borissmidt** [comment]
Maybe with the `export` feature in dotty you could reexpose some of the methods again. But i haven't tried either.

**Pull request merged**